### PR TITLE
work with symbolized strings

### PR DESCRIPTION
**Pull request by bastelfreak**

### CODE REVIEW

File: spec/rspec_puppet_facts_spec.rb [hunk 1]
```diff
@@ -741,88 +744,33 @@
...
       end
 
-      it 'returns facts from a facter version matching 3.3' do
-        is_expected.to match(
-          'centos-7-x86_64' => include(:facterversion => '3.3.0'),
-        )
-      end
-    end
-
-    context 'With a custom facterversion (3.3.2) in the options hash' do
```
Comment (bastelfreak) [inline spec/rspec_puppet_facts_spec.rb:795]: those tests were redundant so I dropped them

**Comment by bastelfreak** [comment]
This contains #176

**Review by silug — APPROVED** [review]

File: lib/rspec-puppet-facts.rb [hunk 2]
```diff
@@ -148,26 +148,16 @@ def on_supported_os_implementation(opts = {})
...
       else
-        if facts[:operatingsystemmajrelease].nil?
-          operatingsystemmajrelease = facts[:operatingsystemrelease].split('.')[0]
-        else
-          operatingsystemmajrelease = facts[:operatingsystemmajrelease]
-        end
+        operatingsystemmajrelease = facts[:os]['release']['major']
+        operatingsystem = facts[:os]['name'].downcase
+        hardwaremodel = facts[:os]['hardware']
       end
```
Comment (ekohl) [inline lib/rspec-puppet-facts.rb:159]: How about
```ruby
os = RSpec.configuration.facterdb_string_keys ? facts[:os] : facts['os']
operatingsystemmajrelease = os['release']['major']
operatingsystem = os['name'].downcase
hardwaremodel = os['hardware']
```

File: lib/rspec-puppet-facts.rb [hunk 3]
```diff
@@ -148,26 +148,16 @@ def on_supported_os_implementation(opts = {})
...
-      os = "#{facts[:operatingsystem].downcase}-#{operatingsystemmajrelease}-#{facts[:hardwaremodel]}"
+      os = "#{operatingsystem}-#{operatingsystemmajrelease}-#{hardwaremodel}"
       next unless os.start_with? RspecPuppetFacts.spec_facts_os_filter if RspecPuppetFacts.spec_facts_os_filter
```
Comment (ekohl) [inline lib/rspec-puppet-facts.rb:161]: I never looked at this line before, but I didn't know you could chain `unless` and `if` on a single line. I'd probably write it as a single statement, but not in this patch.
```suggestion
      next if RspecPuppetFacts.spec_facts_os_filter && !os.start_with?(RspecPuppetFacts.spec_facts_os_filter)
```

File: lib/rspec-puppet-facts.rb [hunk 4]
```diff
@@ -148,26 +148,16 @@ def on_supported_os_implementation(opts = {})
...
       facts.merge! RspecPuppetFacts.common_facts
       os_facts_hash[os] = RspecPuppetFacts.with_custom_facts(os, facts)
```
Comment (ekohl) [inline lib/rspec-puppet-facts.rb:163]: Just below is a call to `stringify_keys`, which I think is redundant if you ask FacterDB to give you strings.

**Review by ekohl — COMMENTED** [review]
I'd be tempted to remove all symbol keys from facterdb and then do the `.to_sym` here in `rspec-puppet-facts` if requested. Any thoughts on that?

**Comment by bastelfreak** [comment]
> I'd be tempted to remove all symbol keys from FacterDB and then do the .to_sym here in rspec-puppet-facts if requested. Any thoughts on that?

I considered that, but that would be another breaking change. I think with the current implementation it allows us to release it as an enhancement and then we can use the new rspec-puppet-facts + FacterDB version. Afterwards I would like to update FacterDB as you mentioned.

Basically:

* merge https://github.com/voxpupuli/facterdb/pull/364
* Release FacterDB 2.1
* Merge #176 + #175 
* Release rspec-puppet-facts
* Drop legacy facts in FacterDB
* Switch to strings by default in FacterDB
* Release FacterDB 3
* Update rspec-puppet-facts

File: lib/rspec-puppet-facts.rb (hunk 2 above)
Comment (bastelfreak) [reply lib/rspec-puppet-facts.rb:159]: oh that's way cleaner, I will implement it.

File: lib/rspec-puppet-facts.rb (hunk 4 above)
Comment (bastelfreak) [reply lib/rspec-puppet-facts.rb:158]: yes, done

File: lib/rspec-puppet-facts.rb (hunk 3 above)
Comment (bastelfreak) [reply lib/rspec-puppet-facts.rb:156]: https://github.com/voxpupuli/rspec-puppet-facts/pull/177

File: lib/rspec-puppet-facts.rb [hunk 5]
```diff
@@ -148,33 +148,16 @@ def on_supported_os_implementation(opts = {})
...
-          operatingsystemmajrelease = facts[:operatingsystemrelease].split('.')[0]
-        else
-          operatingsystemmajrelease = facts[:operatingsystemmajrelease]
-        end
-      end
-      os = "#{facts[:operatingsystem].downcase}-#{operatingsystemmajrelease}-#{facts[:hardwaremodel]}"
+      os = RSpec.configuration.facterdb_string_keys ? facts['os'] : facts[:os]
+      operatingsystemmajrelease = os['release']['major']
+      operatingsystem = os['name'].downcase
+      hardwaremodel = os['hardware']
```
Comment (bastelfreak) [inline lib/rspec-puppet-facts.rb:154]: https://github.com/voxpupuli/facterdb/pull/367 proove that this works always

**Review by ekohl — COMMENTED** [review]
I was further looking at the changes and thought that moving to modern facts deserved its own PR for a cleaner history. https://github.com/voxpupuli/rspec-puppet-facts/pull/178 does that. I'd prefer to rebase this PR on top. As you can see, I already introduce the `os_fact` variable there, so it should be easy to modify.

File: spec/rspec_puppet_facts_spec.rb [hunk 6]
```diff
@@ -855,8 +816,8 @@
             {
               "operatingsystem" => "RedHat",
               "operatingsystemrelease" => [
-                "6",
-                "7",
+                "8",
+                "9"
```
Comment (ekohl) [inline spec/rspec_puppet_facts_spec.rb:820]: Mind keeping the trailing comma?
```suggestion
                "9",
```

**Review by ekohl — APPROVED** [review]

**Comment by ekohl** [comment]
Did you intend to squash all your changes in a single commit?

**Pull request merged**